cmd/titus-storage/ebs.go

```diff
@@ -0,0 +1,272 @@
+	return true
+}
+
+func doesDriveExistWithTimeout(driveName string, maxAttempts int) bool {
+	var attempts int
+	for !doesDriveExist(driveName) {
+		time.Sleep(2 * time.Second)
+		attempts++
+		if attempts >= maxAttempts {
+			return false
+		}
+	}
+	return true
+}
+
+// GetActualBlockDeviceName returns the actual name of the block device seen
+// within the instance, like /dev/nvme1
+// This is often different from the ec2 device name, which is usually
+// something like /dev/sdx
+func GetActualBlockDeviceName(name string) (string, error) {
+	for _, device := range listNvmeBlockDevices() {
+		if d, _ := ebsnvme.ScanDevice(device); d.Name == name {
+			return device, nil
+		}
+	}
+	if _, err := os.Stat(name); os.IsNotExist(err) {
+		return "", err
+	}
+	return name, nil
+}
+
+func listNvmeBlockDevices() (devices []string) {
+	re := regexp.MustCompile(`(^\/dev\/nvme[0-9]+n1$)`)
+	f, _ := filepath.Glob("/dev/nvme*")
+	for _, d := range f {
+		if re.Match([]byte(d)) {
+			devices = append(devices, d)
+		}
+	}
+	return
+}
+
+// getDeviceName scans the Nvme on disk and looks at the
+// serial numbers of the devices and returns the device name
+// (/dev/nmve1). Behaves like a hash get, returning key, ok.
+// returning "", false if unfound.
+func getDeviceName(ebsVolumeID string) (string, bool) {
+	deviceName := ""
+	for _, device := range listNvmeBlockDevices() {
+		if d, _ := ebsnvme.ScanDevice(device); d.VolumeID == ebsVolumeID {
+			deviceName = d.Name
+		}
+	}
+	return deviceName, deviceName != ""
+}
+
+func attachEBS(ebsVolumeID string, ec2Client *ec2.EC2, instanceID string) (string, error) {
+	deviceName, ok := getDeviceName(ebsVolumeID)
+	if ok {
+		log.Printf("EBS Volume %s already attached as %s", ebsVolumeID, deviceName)
+		return deviceName, nil
+	}
+	deviceName, err := randDriveNamePicker()
+	if err != nil {
+		log.Fatal("Couldn't find an unused drive name")
+	}
+	attachVolIn := &ec2.AttachVolumeInput{
+		Device:     &deviceName,
+		InstanceId: &instanceID,
+		VolumeId:   &ebsVolumeID,
+	}
+	volAttachments, err := ec2Client.AttachVolume(attachVolIn)
+	if err != nil {
+		log.Fatalf("Couldn't attach: %v", err)
+	}
+	log.Printf("%+v", volAttachments)
+
+	if !doesDriveExistWithTimeout(deviceName, 10) {
+		log.Fatalf("Drive %s (fake ec2 device name) doesn't exist for volume %s after attaching - checked with stat 10 times", deviceName, ebsVolumeID)
+	}
+	return deviceName, nil
+}
+
+func detatchEBS(ebsVolumeID string, ec2Client *ec2.EC2, instanceID string) error {
+	deviceName, ok := getDeviceName(ebsVolumeID)
+	if !ok {
+		log.Printf("volume %s doesn't seem to connected. Not bothering to detach using the AWS API", ebsVolumeID)
+	}
+	detachVolIn := &ec2.DetachVolumeInput{
+		Device:     &deviceName,
+		InstanceId: &instanceID,
+		VolumeId:   &ebsVolumeID,
+	}
+	volAttachments, err := ec2Client.DetachVolume(detachVolIn)
+	if err != nil {
+		log.Fatal(err)
```

Comment:
Maybe return error here, there might be cleanup needed in caller?

Reply:
I'm wondering (and this is an open question to you all) how we feel about using `log.Fatal()` all over the place?  It seems like if we ever want to write unit tests for this code (and maybe we don't), that returning errors and then having `main()` do the `log.Fatal()` might make that easier.

Reply:
I agree. I'm going to consolidate all the fatals into returns and let the main runner handle the actual exits